src/main/resources/jupyter/edit-mode.js

```diff
@@ -0,0 +1,337 @@
+            .then(res => {
+                handleLockStatus(res)
+                return res.json()
+            })
+            .catch(err => {
+                console.error(err)
+            })
+    }
+
+    async function toggleMetaFailureBanner(shouldShow) {
+        const bannerId = "notification_metaFailure"
+        const bannerText = "Failed to check notebook status, changes may not be saved to workspace. Retrying..."
+
+        removeElementById(bannerId)
+
+        if (shouldShow) {
+            const bannerStyling = "btn btn-xs navbar-btn btn-danger"
+
+            $('#notification_area').append(
+                $('<div>').attr({
+                    "id": bannerId,
+                    "class": bannerStyling
+                })
+                .append($('<span>').html('<i class="fa fa-exclamation-triangle"></i> ' + bannerText))
+            );
+        }
+
+    }
+
+    async function handleLockStatus(res) {
+        if (!res.ok) {
+            const status = res.status
+            const errorText = res.statusText
+
+            if (status == 409) {
+                res.json().then(body => {
+                    const message = getLockConflictBody(body.lockedBy)
+                    promptUserWithModal(lockConflictTitle, lockIssueButtons, message)
+                })
+            }
+            //for the lock endpoint, we consider all non 'ok' statuses an error
+            throw new Error(errorText)
+        }
+    }
+
+    const getLockConflictBody = (lockHolder) => {
+        return `<p>This file is currently being editted by ${lockHolder}.</p>` +
+            `<br/><p>You can make a copy, or run it in Playground Mode to explore and execute its contents without saving any changes.`;
+    }
+
+    async function promptUserWithModal(title, buttons, htmlBody) {
+        if (modalOpen) return
+
+        modalOpen = true
+
+        dialog.modal({
+                body: $('<p>').html(htmlBody),
+                title: title,
+                buttons: buttons,
+                notebook: Jupyter.notebook,
+                keyboard_manager: Jupyter.notebook.keyboard_manager
+            })
+            .on('hidden.bs.modal', () => modalOpen = false)
+            .attr('id', 'leoUserModal')
+            .find(".close").click(() => window.history.back()) //TODO: test in docker image
+    }
+
+    async function openPlaygroundMode() {
+        const url = jupyterContentsAPIUrl + Jupyter.notebook.notebook_path
+        const newPath = meta.storageLink.localSafeModeBaseDirectory.replace('notebooks/', '') + '/' + Jupyter.notebook.notebook_name
+
+        console.log('switching to playground path: ', newPath)
+
+        const payload = {
+            ...basePayload,
+            method: 'PATCH',
+            body: JSON.stringify({ path: newPath })
+        }
+
+        console.log('url: ', url, newPath)
+
+        fetch(url, payload).then(res => {
+            window.location.href = jupyterBaseUrl + "notebooks/" + newPath
```

Comment:
You should be able to simplify all of this just using relative paths, e.g. `window.location.href = "/notebooks/" + newPath`

If there's a basePath set, you could even go potentially simpler: `window.location.href = newPath`

Reply:
As I understand it, we will need to call the leo jupyter proxy endpoint (jupyterBaseUrl), which will then proxy the call to the Jupyter server at the API path specified after the leo path. I had this working locally on my Jupyter server, but this will need more testing once I have the extension in the jupyter image. 

That being said, I didn't know you could specify a relative path when assigning window.location.href, so I'll investigate simplifying this